mu-plugins/enforce-2fa.php

```diff
@@ -0,0 +1,438 @@
+
+// Set the cookies when the user logs in.
+add_action(
+	'jetpack_sso_handle_login',
+	function ( $user, $user_data ) {
+
+		add_action(
+			'set_auth_cookie',
+			function ( $auth_cookie, $expire, $expiration, $user_id, $scheme, $token ) use ( $user_data ) {
+
+				// Check if the site is secure.
+				$secure = is_ssl();
+
+				// Set the cookies for Jetpack SSO.
+				$sso_cookie = wp_generate_auth_cookie( $user_id, $expire, 'secure_auth', $token );
+				setcookie( E2FA_IS_JETPACK_SSO_COOKIE, $sso_cookie, $expire, COOKIEPATH, COOKIE_DOMAIN, $secure, true );
+
+				// Set the cookies for Jetpack SSO 2SA.
+				if ( $user_data->two_step_enabled ) {
+
+					// Set the cookies for Jetpack SSO 2SA.
+					$sso_2sa_cookie = wp_generate_auth_cookie( $user_id, $expire, 'secure_auth', $token );
+					setcookie( E2FA_IS_JETPACK_SSO_2SA_COOKIE, $sso_2sa_cookie, $expire, COOKIEPATH, COOKIE_DOMAIN, $secure, true );
+				}
+			},
+			10,
+			6
+		);
+	},
+	10,
+	2
+);
+
+// Clear the cookies when the user logs out.
+add_action(
+	'clear_auth_cookie',
+	function () {
+
+		if ( ! headers_sent() ) {
+
+			// Clear the cookies for Jetpack SSO.
+			setcookie( E2FA_IS_JETPACK_SSO_COOKIE, ' ', time() - YEAR_IN_SECONDS, COOKIEPATH, COOKIE_DOMAIN );
+
+			// Clear the cookies for Jetpack SSO 2SA.
+			setcookie( E2FA_IS_JETPACK_SSO_2SA_COOKIE, ' ', time() - YEAR_IN_SECONDS, COOKIEPATH, COOKIE_DOMAIN );
+		}
+	}
+);
+
+/**
+ * Check if the user is logged in with Jetpack SSO.
+ *
+ * @return bool
+ */
+function e2fa_is_jetpack_sso() {
+
+	// If the user is not logged in, return early.
+	if ( ! is_user_logged_in() ) {
+		return false;
+	}
+
+	// If the user is not logged in with Jetpack SSO, return early.
+	if ( ! isset( $_COOKIE[ E2FA_IS_JETPACK_SSO_COOKIE ] ) ) {
+		return false;
+	}
+
+	// Check if the user is logged in with Jetpack SSO.
+	$cookie = sanitize_text_field( wp_unslash( $_COOKIE[ E2FA_IS_JETPACK_SSO_COOKIE ] ) );
+	return wp_validate_auth_cookie( $cookie, 'secure_auth' );
+}
+
+/**
+ * Check if the user is logged in with Jetpack SSO and has 2SA enabled.
+ *
+ * @return bool
+ */
+function e2fa_is_jetpack_sso_two_step() {
+
+	// If the user is not logged in, return early.
+	if ( ! e2fa_is_jetpack_sso() ) {
+		return false;
+	}
+
+	// If the user is not logged in with Jetpack SSO 2SA, return early.
+	if ( ! isset( $_COOKIE[ E2FA_IS_JETPACK_SSO_2SA_COOKIE ] ) ) {
+		return false;
+	}
+
+	// Check if the user is logged in with Jetpack SSO 2SA.
+	$cookie = sanitize_text_field( wp_unslash( $_COOKIE[ E2FA_IS_JETPACK_SSO_2SA_COOKIE ] ) );
+	return wp_validate_auth_cookie( $cookie, 'secure_auth' );
+}
+
+/**
+ * Should Force 2FA.
+ *
+ * @return bool
+ */
+function e2fa_should_force_two_factor() {
+
+	// Fallout, if Localhost.
+	if ( 'local' === wp_get_environment_type() || ( isset( $_SERVER['HTTP_HOST'] ) && in_array( $_SERVER['HTTP_HOST'], array( 'localhost', '127.0.0.1', '[::1]', '192.168.0.1', '192.168.1.1' ), true ) ) ) {
```

Comment:
Can you please confirm if `$_SERVER['HTTP_HOST']` always returns `localhost` if the request URI is `127.0.0.1` or `[::1]`? AFAIK it does by default.

Reply:
I quite don't understand this comment.